src/Datagrid/ProxyQuery.php

```diff
@@ -211,16 +212,7 @@ public function execute(array $params = [], $hydrationMode = null)
             return $query->execute($params, $hydrationMode);
         }
 
-        $identifierFieldNames = $this
-            ->getQueryBuilder()
-            ->getEntityManager()
-            ->getMetadataFactory()
-            ->getMetadataFor(current($this->getQueryBuilder()->getRootEntities()))
-            ->getIdentifierFieldNames();
-
-        // Paginator with fetchJoinCollection doesn't work with composite primary keys
-        // https://github.com/doctrine/orm/issues/2910
-        return new Paginator($query, 1 === \count($identifierFieldNames));
+        return SmartPaginatorFactory::create($this, $this->hints);
```

Comment:
At first I thought the `QueryBuilder` was enough, but we need the query from `getDoctrineQuery` and also `QueryBuilder`.